### PR TITLE
Created React component for pagination

### LINKED ISSUE #595: Refactor pagination widget into React component
This is a prerequisite toward getting pagination to trigger AJAX calls

### PR DESCRIPTION
**Pull request by noisecapella**
Fixes #595

### CODE REVIEW

File: ui/static/ui/js/listing.js [hunk 1]
```diff
@@ -355,12 +279,14 @@ define('listing',
...
-      $("a#repo_page_status").click(function() {
-        var pageNum = $(this).data('page-num');
-        var maxPages = $(this).data('max-page-num');
-        navigateToPage(pageNum, maxPages, this);
-      });
+      var updatePage = function(newPageNum) {
+        var urlMap = URI(window.location).query(true);
+        urlMap.page = newPageNum;
+        window.location = "?" + URI().search(urlMap).query();
+      };
```
Comment (giocalitri) [inline ui/static/ui/js/listing.js:331]: Question: does this mean that we are tight to the necessary informations being in the URL?

Comment (noisecapella) [inline ui/static/ui/js/listing.js:331]: The `pageNum` is passed in the `loader` function above, we don't read it in the URL. This just sets the window location with the updated page parameter. In the near future we can update this callback to just change `pageNum` and call the API with the new page number.

Comment (giocalitri) [inline ui/static/ui/js/listing.js:331]: got it

File: ui/static/ui/js/pagination.jsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  var Pagination = React.createClass({
+    mixins: [React.addons.LinkedStateMixin],
+    getInitialState: function() {
+      return {
+        isEditing: false,
+        textPageNum: this.props.pageNum
+      };
+    },
+    render: function() {
+      if (this.props.numPages <= 1) {
```
Comment (giocalitri) [inline ui/static/ui/js/pagination.jsx:13]: minor: maybe check that it is an integer >=1?

File: ui/static/ui/js/pagination.jsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      var $node = $(React.findDOMNode(component));
+      $node.focus();
+      var thiz = this;
+      $node.focusout(function() {
+        thiz.goTo(thiz.state.textPageNum);
+      });
+    },
+    goToPrevious: function(e) {
+      e.preventDefault();
+      this.goTo(this.props.pageNum - 1);
```
Comment (giocalitri) [inline ui/static/ui/js/pagination.jsx:64]: extra check that `this.props.pageNum` is >1?

Comment (giocalitri) [inline ui/static/ui/js/pagination.jsx:64]: maybe is not necessary if you don't allow to click on the "previous" button

File: ui/static/ui/js/require_config.js [hunk 4]
```diff
@@ -9,6 +9,10 @@ var REQUIRE_PATHS = {
   react_infinite: "react-infinite/dist/react-infinite",
   lodash: "lodash/lodash",
   select2: "select2/dist/js/select2.full",
+  uri: "uri.js/src/URI",
+  punycode: "uri.js/src/punycode",
+  IPv6: "uri.js/src/IPv6",
+  SecondLevelDomains: "uri.js/src/SecondLevelDomains",
```
Comment (giocalitri) [inline ui/static/ui/js/require_config.js:15]: question: why all these requirements? dependencies?

**Comment by giocalitri** [comment]
Some questions, but overall the code looks good

**Comment by giocalitri** [comment]
Functionality looks fine as well

File: ui/static/ui/js/pagination.jsx (hunk 2 above)
Comment (noisecapella) [inline ui/static/ui/js/pagination.jsx:13]: What are you suggesting, can you clarify?

File: ui/static/ui/js/require_config.js (hunk 4 above)
Comment (noisecapella) [inline ui/static/ui/js/require_config.js:15]: Yes, the other three are required by `URI.js` and `URI.js` is used to parse the URL query string

File: ui/static/ui/js/pagination.jsx (hunk 2 above)
Comment (giocalitri) [inline ui/static/ui/js/pagination.jsx:13]: I was suggesting to check that `this.props.numPages` is an integer and not float
but I checked that it does not create weird problem (it gives a 404), so feel free to not consider this comment

**Comment by giocalitri** [comment]
I guess this can be merged

**Pull request merged**